### ISSUE TITLE
Add option to reduce external surface of predefined sections by a corrosion allowance

### ISSUE THREAD

**Issue by Agent6-6-6** [robbievanleeuwen/section-properties#9]
**Is your feature request related to a problem? Please describe.**
Often for steel in piles or marine environments you are required to assess the section properties of a section after some defined section loss from the perimeter. Like say applying an annual corrosion loss from exposed surfaces multiplied by a number of years. 

**Describe the solution you'd like**
A means of taking an original section and an additional parameter of a surface thickness loss and applying this loss of section to the external perimeter before creating the geometry and mesh.
 
For an I section for example altering within a pre-processing procedure that reduces the section inputs as follows (C = surface thickness loss) prior to creating the geometry:-
```
d = d - 2C
b = b - 2C
t_f = t_f - 2C
t_w = t_w - 2C
r = r + 2C
```
etc... for other shapes
Fairly easy to do in your own pre-processing code, but though it might be a nice addition to your programme for in built sections at least (and may even be able to be generalised for custom shapes). Would allow for looping through the properties over a timeframe to plot loss of capacity for example, useful for determining remaining life perhaps (may make a good advanced example similar to the others that you currently have).

Could also allow you to plot corroded section geometry plot against original on same plot (perhaps different coloured line or something)?

**Describe alternatives you've considered**
Do it in my own custom pre-processing code, adjusting section geometry inputs before creating reduced section geometry.

**Comment by robbievanleeuwen** [comment]
Included this feature in cd22233.

Have done a few tests and it seems to work as expected. Agent6-6-6, if you have any test cases would appreciate if you could have a quick look at this to see what you think. After I work on incorporating the shapely package more within the program this feature will become more robust, but keen to hear your original thoughts.

Cheers,
Robbie

**Comment by Agent6-6-6** [comment]
I'll take a look when I get a chance and dig up the specific case I looked at this on to compare answers. Probably be next weekend before I get a chance though.

Regards.

**Comment by robbievanleeuwen** [comment]
Perfect! Took me a year to get around to doing this so all good :)

**Comment by Agent6-6-6** [comment]
Hi Robbie

Managed to check a couple of things with the following comments: -

By comparing a section geometry generated with manually applying a surface offset to the inputs, and using the new offset function I notice a small difference in geometry. The geometry generated seems to be slightly different, when I would have thought they would have been the same?

I think it's due to the way the offsets work on the radii in sections with shapely, because it's not a true curve but a series of short lines. If you consider the point where a curve meets a straight portion of the section in say an RHS, and a line back to the origin of the radii, if this point is offset it should still remain on the same line. However it offsets at an angle midway between the angle between the two lines. This effect is lessened with more points in the curve. I don't think there is any way around this but it does introduce some error compared with offsetting a true curve in say AutoCAD.

You can see the effect in this screenshot with a small number of segments around the curve, you'll note that the portion around the curve is noticably thinner?

![Figure_1](https://user-images.githubusercontent.com/36808164/82139492-39b8cb00-987c-11ea-8688-2f2ad3d3c7ad.png)

**Comment by Agent6-6-6** [comment]
I think you can also use buffer in shapely package, this might remove the need to specify left or right for the offset which may simplify inputs required?

**Comment by robbievanleeuwen** [comment]
Hi @Agent6-6-6, picking this one up again. What are your thoughts on the current `offset_perimeter` method in the v2 code base? Does it perform as expected/is there any scope to improve/build upon the method?

**Comment by Agent6-6-6** [comment]
@robbievanleeuwen, I'll try take a look during the week. I think from memory @connorferster implemented the same thing but with option to do internal, external perimeters, or both using shapely. 

Been a while since I've needed/used this option.

**Comment by connorferster** [comment]
@Agent6-6-6 (& @robbievanleeuwen ),

I do believe I wrote the offset_perimeter method with Agent6-6-6's feedback to include the interior and/or exterior erosion/dilation separately to handle this scenario. However, for plotting both reduced and original section's together, there is not "built-in" method for doing so: the user would have to build their own plots from using the returned (fig, ax) from two separate plots.

**Comment by robbievanleeuwen** [comment]
Ok sounds good! If all is good will be looking to close this one :)

**Comment by robbievanleeuwen** [comment]
Note - check this works with `CompoundGeometry` and also check perimeter calculation.

**Comment by robbievanleeuwen** [comment]
Implemented with #140
